Implement basic auth and limit stream hooks to admins only

victorges opened the pull request:
<!-------------------------------------------------------------------------
 | Thanks for send a pull request! 🎉
 | First, please make sure you familiar with the contribution guidelines
 | https://github.com/livepeer/livepeer.com/blob/master/CONTRIBUTING.md
 -------------------------------------------------------------------------->

**What does this pull request do? Explain your changes. (required)**
This is to add an authorization layer to the hooks called on the stream object by go-livepeer.

Solution has been discussed in the issue: https://github.com/livepeer/go-livepeer/issues/1278

**Specific updates (required)**
 - Add support to HTTP Basic Auth in auth middleware
 - Make stream auth and detection hooks admin-only
 - TODO: Add an admin token to go-livepeer deployments for them to call the API

## -

- **How did you test each of these updates (required)**
 - For now mostly `yarn test` which already got some errors on the tests calling the hooks with a non admin token
 - TODO: Create some tests for the Basic auth logic

**Does this pull request close any open issues?**
Fixes https://github.com/livepeer/go-livepeer/issues/1278

**Checklist:**

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have read the **CONTRIBUTING** document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.

victorges committed:
Create tokenType temp on authorization middleware

victorges committed:
Add basic-auth support on auth middleware

victorges committed:
Use db instead of store on auth mware

victorges committed:
Only allow admins to call the stream hooks

victorges committed:
Fix tests :sparkles:

iameli reviewed (approved):
Love it. But let's wait to merge until the go-livepeer bits land, yes?

darkdarkdragon commented on packages/api/src/controllers/stream.js:
```diff
@@ -1233,7 +1233,10 @@ async function terminateStreamReq(req, stream) {
 
-app.post("/hook", async (req, res) => {
+app.post("/hook", authMiddleware({}), async (req, res) => {
+  if (!req.user.admin && !req.user.access?.broadcasterHooks) {
```
Can we move this into `auth-middleware`, similar to `anyAdmin` flag?

victorges replied:
The problem with that would be how to represent the "admin OR broadcasterHooks" idea in the params object. For example something like `{ anyAdmin: true, access: { broadcasterHooks: true } }` could make sense for this specific case, but it could also look like that _both_ things need to be set instead, and not one or another. 

I could make a more generic interface in the auth middleware, like receiving a function that validates properties on the `User` object (and token type?) and returns a `boolean` instead. Then already provide some helper implementations like `isAdmin`, `isAnyAdmin` and `isAnyAdminOrBroadcaster` which can be set on the params objects. WDYT?

victorges replied:
Implemented the idea we agreed offline :)

darkdarkdragon commented on packages/api/src/middleware/auth.js:
```diff
@@ -11,31 +19,37 @@ function authFactory(params) {
+      throw new ForbiddenError(`no authorization header provided`);
+    } else if (["Bearer", "Basic"].includes(tokenType)) {
+      const isBasic = tokenType === "Basic";
+      tokenObject = await db.apiToken.get(isBasic ? basicUser.pass : req.token);
```
`basicUser.pass ` - this assumes that `basicUser` was successfully parsed, which not necessary be the case.

victorges replied:
Good catch. I could swear I read somewhere it returned an empty object when the parsing failed, but apparently it's not the case.

darkdarkdragon replied:
Empty object will not work too - then `db.apiToken.get` will throw error because it expect id to be provided.

victorges replied:
I see. Alright so I can't even coalesce to an empty string here, I'll do an explicit check.

victorges replied:
While we're at it, do you know where this `req.token` property comes from? I coudn't find it on express docs nor any code of ours setting it explicitly.

victorges replied:
Nevermind just found it: https://github.com/livepeer/livepeer-com/blob/d903f4e5761b05bd7469c925c8a4201f558ca38e/packages/api/src/app-router.js#L120

Maybe I'll use the [`express-basic-auth`](https://www.npmjs.com/package/express-basic-auth) instead of this more manual `basic-auth` then, for consistency.

victorges replied:
Fixed. I'll check again later if it's worth migrating to the `express-basic-auth`, it seems to do way more than just parsing the headers so it may be overkill.

victorges committed:
Check if basic auth was actually parsed

victorges committed:
Fix admin/uiadmin validation

We weren't disallowing basic auth from being treated
as a UI admin, when it should not. Changed the code to
actually reuse the isUIAdmin boolean that was already
calculated to set in the request.

darkdarkdragon reviewed (changes requested):
1. Can we remove last usage of `req.store` from `auth.js`?
2. Can we fix tests?

victorges committed:
Remove last req.store and fix tests

darkdarkdragon reviewed (approved):
LGTM

iameli commented on packages/api/src/middleware/auth.js:
```diff
@@ -11,74 +19,70 @@ function authFactory(params) {
+      tokenObject = await db.apiToken.get(tokenId);
+      const matchesBasicUser = tokenObject?.userId === basicUser?.name;
+      if (!tokenObject || (isBasic && !matchesBasicUser)) {
+        throw new ForbiddenError(`no token object ${tokenValue} found`);
```
Nice, this will finally get rid of the really awkward `no token object undefined found`

iameli reviewed (approved):
LGTM also. Let's make the go-livepeer changes and merge at the start of the load test session?

victorges committed:
Add some basic tests :ba-dum-tss:
